src/search/dag_classic/search.cc

```diff
@@ -0,0 +1,2418 @@
+      if (n > 0 && q > -0.98f) {
+        stats->may_resign = false;
+      }
+      if (max_n < n) {
+        max_n = n;
+        max_n_has_max_q_plus_m = false;
+      }
+      if (max_q_plus_m <= q_plus_m) {
+        max_n_has_max_q_plus_m = (max_n == n);
+        max_q_plus_m = q_plus_m;
+      }
+    }
+    if (!max_n_has_max_q_plus_m) {
+      stats->time_usage_hint_ = IterationStats::TimeUsageHint::kNeedMoreTime;
+    }
+  }
+}
+
+void Search::WatchdogThread() {
+  LOGFILE << "Start a watchdog thread.";
+  StoppersHints hints;
+  IterationStats stats;
+  while (true) {
+    PopulateCommonIterationStats(&stats);
+    MaybeTriggerStop(stats, &hints);
+    MaybeOutputInfo();
+
+    constexpr auto kMaxWaitTimeMs = 100;
+    constexpr auto kMinWaitTimeMs = 1;
+
+    Mutex::Lock lock(counters_mutex_);
+    // Only exit when bestmove is responded. It may happen that search threads
+    // already all exited, and we need at least one thread that can do that.
+    if (bestmove_is_sent_) break;
+
+    auto remaining_time = hints.GetEstimatedRemainingTimeMs();
+    if (remaining_time > kMaxWaitTimeMs) remaining_time = kMaxWaitTimeMs;
+    if (remaining_time < kMinWaitTimeMs) remaining_time = kMinWaitTimeMs;
+    // There is no real need to have max wait time, and sometimes it's fine
+    // to wait without timeout at all (e.g. in `go nodes` mode), but we
+    // still limit wait time for exotic cases like when pc goes to sleep
+    // mode during thinking.
+    // Minimum wait time is there to prevent busy wait and other threads
+    // starvation.
+    watchdog_cv_.wait_for(
+        lock.get_raw(), std::chrono::milliseconds(remaining_time),
+        [this]() { return stop_.load(std::memory_order_acquire); });
+  }
+  LOGFILE << "End a watchdog thread.";
+}
+
+void Search::FireStopInternal() {
+  stop_.store(true, std::memory_order_release);
+  watchdog_cv_.notify_all();
+}
+
+void Search::Stop() {
+  Mutex::Lock lock(counters_mutex_);
+  ok_to_respond_bestmove_ = true;
+  FireStopInternal();
+  LOGFILE << "Stopping search due to `stop` uci command.";
+}
+
+void Search::Abort() {
+  Mutex::Lock lock(counters_mutex_);
+  if (!stop_.load(std::memory_order_acquire) ||
+      (!bestmove_is_sent_ && !ok_to_respond_bestmove_)) {
+    bestmove_is_sent_ = true;
+    FireStopInternal();
+  }
+  LOGFILE << "Aborting search, if it is still active.";
+}
+
+void Search::Wait() {
+  Mutex::Lock lock(threads_mutex_);
+  while (!threads_.empty()) {
+    threads_.back().join();
+    threads_.pop_back();
+  }
+}
+
+void Search::CancelSharedCollisions() REQUIRES(nodes_mutex_) {
+  for (auto& entry : shared_collisions_) {
+    auto path = entry.first;
+    for (auto it = ++(path.crbegin()); it != path.crend(); ++it) {
+      std::get<0>(*it)->CancelScoreUpdate(entry.second);
+    }
+  }
+  shared_collisions_.clear();
+}
+
+Search::~Search() {
+  Abort();
+  Wait();
+  {
+    SharedMutex::Lock lock(nodes_mutex_);
+    CancelSharedCollisions();
+
+#ifndef NDEBUG
```

Comment:
assert() is already a noop when `NDEBUG` is set, it's not need to guard it here.